### PR TITLE
Add support for using Dispatch in lit tests on Linux

### PR DESCRIPTION
**Pull request by kubamracek**
Today, several lit tests use and test Dispatch, but this is only available on macOS/iOS and gated behind `REQUIRES: objc_interop`. This PR introduces a different flag, `REQUIRES: dispatch` and makes it available on Linux (if you use build-script with --libdispatch). The individual pieces:

- In build-script, when building libdispatch on Linux, copy the resulting libdispatch.so and swiftmodule files to the right location in the Swift resource directory.
- In lit, detect whether the build has libdispatch.so, and add `dispatch` as an available feature. On macOS, this is always available.
- Change a few tests to use `REQUIRES: dispatch` instead of `REQUIRES: objc_interop`
- Remove the Dispatch module from mock SDKs.

### CODE REVIEW

**Comment by kubamracek** [comment]
@swift-ci please test

File: test/lit.cfg [hunk 1]
```diff
@@ -813,6 +814,9 @@ elif run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'windows-cygnus', 'wi
       config.target_object_format = "elf"
       config.target_dylib_extension = "so"
       config.target_sdk_name = "linux"
+      if os.path.exists(make_path(config.swift_lib_dir, "swift", "linux", "libdispatch.so")):
```
Comment (compnerd) [inline test/lit.cfg]: Do we ever test with the static variant of the runtime?  We may have those enabled, so we need to check both :-(.

File: test/stdlib/DispatchData.swift [hunk 2]
```diff
@@ -23,7 +25,13 @@ DispatchAPI.test("dispatch_data_t deallocator") {
 	let q = DispatchQueue(label: "dealloc queue")
 	var t = 0
 
-	autoreleasepool {
+#if os(macOS) || os(iOS)
+	let runWithPool: (()->())->() = { autoreleasepool { $0() } }
+#else
+	let runWithPool: (()->())->() = { $0() }
+#endif
```
Comment (compnerd) [inline test/stdlib/DispatchData.swift]: OMG, this is beautiful!  I ended up doing some more ... intensive refactoring in swift-syntax to deal with the autoreleasepool.

File: utils/build-script-impl [hunk 3]
```diff
@@ -2618,20 +2618,33 @@ for host in "${ALL_HOSTS[@]}"; do
...
+                      cp src/swift/Dispatch.{swiftdoc,swiftmodule} ${SWIFT_BUILD_PATH}/lib/swift/${SWIFT_HOST_VARIANT}/${SWIFT_HOST_VARIANT_ARCH}/
+                  rm -f ${SWIFT_BUILD_PATH}/lib/swift/os
+                  ln -s ${LIBDISPATCH_SOURCE_DIR}/os ${SWIFT_BUILD_PATH}/lib/swift/os
+                  rm -f ${SWIFT_BUILD_PATH}/lib/swift/dispatch
+                  ln -s ${LIBDISPATCH_SOURCE_DIR}/dispatch ${SWIFT_BUILD_PATH}/lib/swift/dispatch
 
-                    -DENABLE_TESTING=YES
-                  )
+                  # libdispatch builds itself
+                  continue
```
Comment (compnerd) [inline utils/build-script-impl]: Woah, I do not understand this magic at all.  Why can we not use the normal build/install rules?

**Review by vlm — APPROVED** [review]

File: test/Inputs/clang-importer-sdk/usr/include/dispatch.h [hunk 4]
```diff
@@ -1,30 +0,0 @@
-#import <objc/NSObject.h>
```
Comment (jrose-apple) [inline test/Inputs/clang-importer-sdk/usr/include/dispatch.h]: This seems suspect to me. test/Inputs/clang-importer-sdk is a fake SDK under the control of the tests, which means anything depending on this should *not* be using the real Dispatch.

**Review by jrose-apple — CHANGES_REQUESTED** [review]

**Pull request closed without merging**